### PR TITLE
feat: 4912 crudable keys

### PR DESCRIPTION
**Pull request by joshbalfour**
# Pull request checklist

**Detail as per issue below (required):**

fixes: #4912

### CODE REVIEW

File: packages/app-builder-backend/src/resolvers/key-resolver.ts [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+  async checkOutKey(
+    @Arg('propertyId') propertyId: string,
+    @Arg('keyId') keyId: string,
+    @Arg('movement', () => KeyMovementModelInput) movement: KeyMovementModelInput,
+    @Ctx() ctx: Context,
+  ): Promise<Key> {
+    const { accessToken, idToken } = ctx
+    if (!accessToken || !idToken) {
+      throw new Error('unauthorized')
+    }
```
Comment (bashleigh) [inline packages/app-builder-backend/src/resolvers/key-resolver.ts:208]: could you make this snippet reusable in any way? Just thinking if we changed the auth method it would take a lot of regex to find all the locations

Comment (bashleigh) [reply packages/app-builder-backend/src/resolvers/key-resolver.ts:208]: Might also be helpful to throw this as a http error that lambdas like? I'm not 100% on this but I know they do like an object with a status property `return { status: 401, message: "Unauthorised" 
 }`

File: packages/app-builder-backend/src/utils/graphqlFetch.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    throw new Error(json.message)
+  }
+  if (json.data) {
+    return json.data[dataKey] as T
+  }
+  if (json.errors) {
+    throw new Error(json.errors[0].message)
+  }
+  throw new Error('Invalid response')
+}
```
Comment (bashleigh) [inline packages/app-builder-backend/src/utils/graphqlFetch.ts:41]: EOF

File: packages/app-builder/src/components/hooks/objects/use-sub-objects.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    const { args } = list
+    const isForThis = args.some((arg) => arg.name === `${lowercaseFirstLetter(typeName)}Id`)
+    const hasOtherRequiredArgs =
+      args.filter((arg) => !(arg.name === `${lowercaseFirstLetter(typeName)}Id` || context[arg.name])).length === 0
+
+    return isForThis && hasOtherRequiredArgs
+  })
+
+  return {
+    data: subobjects,
```
Comment (bashleigh) [inline packages/app-builder/src/components/hooks/objects/use-sub-objects.ts:29]:
```suggestion
  const subObjects = data.filter((type) => {
    const { list } = type
    if (!list) return false
    const { args } = list
    const isForThis = args.some((arg) => arg.name === `${lowercaseFirstLetter(typeName)}Id`)
    const hasOtherRequiredArgs =
      args.filter((arg) => !(arg.name === `${lowercaseFirstLetter(typeName)}Id` || context[arg.name])).length === 0

    return isForThis && hasOtherRequiredArgs
  })

  return {
    data: subObjects,
```

File: packages/app-builder/src/components/hooks/objects/use-sub-objects.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import { useIntrospection } from '../use-introspection'
+import { usePageId } from '../use-page-id'
+
+export const lowercaseFirstLetter = (string: string) => string.charAt(0).toLowerCase() + string.slice(1)
+
+export const useSubObjects = (typeName?: string) => {
```
Comment (bashleigh) [inline packages/app-builder/src/components/hooks/objects/use-sub-objects.ts:6]: Add a return type here? I can see false, {data: T[], error, loading}? and true? so `SubObject | boolean`? Not entrely sure

File: packages/app-builder/src/components/hooks/use-introspection/query-generators.ts [hunk 5]
```diff
@@ -13,37 +13,62 @@ import {
...
+  return {
+    query: gql`
+      ${listQuery}
+    `,
+    args,
+  }
 }
 
-export const getGetQuery = (queries: Array<QueryableField>, queryableObjectTypes: Array<IntrospectionObjectType>) => {
+export const getGetQuery = (
```
Comment (bashleigh) [inline packages/app-builder/src/components/hooks/use-introspection/query-generators.ts:47]: getGetQuery? :p

**Review by bashleigh — COMMENTED** [review]
Only some small issues, the one I'm feeling anal about is the error throws for lambda returns (if they are lambda funcs)

File: packages/app-builder-backend/src/utils/graphqlFetch.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+const getGqlString = (doc: DocumentNode) => doc.loc && doc.loc.source.body
+
+export const query = async <T>(
+  query: DocumentNode,
+  variables: Record<string, any>,
+  dataKey: string,
+  { idToken, accessToken }: AuthTokens,
+) => {
+  const res = await fetch(graphqlUri || 'https://graphql.dev.paas.reapit.cloud/graphql', {
```
Comment (willmcvay) [inline packages/app-builder-backend/src/utils/graphqlFetch.ts:18]: Would be good if this came from the config.json

File: packages/app-builder/src/components/ui/user/ejectable/table.tsx [hunk 7]
```diff
@@ -49,62 +71,122 @@ const DeleteButton = ({ disabled, typeName, id }: { disabled?: boolean; typeName
...
+            label: `View ${uppercaseSentence(object.name)}s`,
+          })),
+        ]
 
-        return {
-          cells: [
-            ...cells,
-            {
-              label: 'Edit',
+        cells = [
```
Comment (willmcvay) [inline packages/app-builder/src/components/ui/user/ejectable/table.tsx:120]: Just a general comment but the table logic is a bit hard to follow. Feels like it could get broken down a bit into some functions or similar. Just a thought though...

**Review by willmcvay — COMMENTED** [review]
Just a couple of small comments from me

File: packages/app-builder-backend/src/resolvers/key-resolver.ts (hunk 1 above)
Comment (joshbalfour) [reply packages/app-builder-backend/src/resolvers/key-resolver.ts:208]: good call, done it the native way now with decorators 👍

File: packages/app-builder/src/components/hooks/use-introspection/query-generators.ts (hunk 5 above)
Comment (joshbalfour) [reply packages/app-builder/src/components/hooks/use-introspection/query-generators.ts:47]: it be like that sometimes 😭

File: packages/app-builder-backend/src/utils/graphqlFetch.ts (hunk 6 above)
Comment (joshbalfour) [reply packages/app-builder-backend/src/utils/graphqlFetch.ts]: it does, i'm just providing a default for types sake

File: packages/app-builder/src/components/ui/user/ejectable/table.tsx [hunk 8]
```diff
@@ -49,26 +71,91 @@ const DeleteButton = ({ disabled, typeName, id }: { disabled?: boolean; typeName
...
+    .map(([label, value]) => ({
+      label: uppercaseSentence(label),
+      value: (typeof value === 'object' ? undefined : value) as string,
+      children: typeof value === 'object' ? <ObjectTableCell obj={value} /> : undefined,
+      narrowTable: {
+        showLabel: true,
+      },
+    }))
+
+const getAdditionalCells = (
```
Comment (willmcvay) [inline packages/app-builder/src/components/ui/user/ejectable/table.tsx:97]: Yeah, much better, easier on the brain!

**Review by willmcvay — APPROVED** [review]
I'm happy if Ashleigh is :-)

**Review by bashleigh — APPROVED** [review]
